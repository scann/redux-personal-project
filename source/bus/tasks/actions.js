//Types
import { types } from './types';

export const tasksActions = {
    //Sync
    fillTasks: (tasks) => {
        return {
            type:    types.FILL_TASKS,
            payload: tasks,
        };
    },
    createTask: (message) => {
        return {
            type:    types.CREATE_TASK,
            payload: message,
        };
    },

    //Async
    fetchTasksAsync: () => {
        return {
            type: types.FETCH_TASKS_ASYNC,
        };
    },
    createTaskAsync: (message) => {
        return {
            type:    types.CREATE_TASK_ASYNC,
            payload: message,
        };
    },
};
