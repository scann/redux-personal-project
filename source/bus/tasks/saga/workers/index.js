export { fetchTasks } from './fetchTasks';
export { createTask } from './createTask';