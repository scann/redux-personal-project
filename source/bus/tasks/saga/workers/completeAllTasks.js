//Core
import { put, select, call } from 'redux-saga/effects';

//Instruments
import { api } from '../../../../REST';
import { tasksActions } from '../../actions';
import { uiActions } from '../../../ui/actions';

export function* completeAllTasks () {
    try {
        yield put(uiActions.startSpinning());
        const tasks = yield select((state) => state.tasks);
        const completedTasks = tasks.map((task) => task.set('completed', true)).toJS();

        const promises = completedTasks.map((task) => api.tasks.complete(task));
        const responses = yield call([Promise, Promise.all], promises);

        for (const response of responses) {
            if (response.status !== 200) {
                throw new Error('Task was not completed');
            }
        }

        yield put(tasksActions.completeAllTasks());
    } catch (error) {
        yield put(uiActions.emitError(error, 'completeAllTasks worker'));
    } finally {
        yield put(uiActions.stopSpinning());
    }
}
