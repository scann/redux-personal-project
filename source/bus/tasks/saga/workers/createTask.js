//Core
import { put, apply } from 'redux-saga/effects';
import { actions } from 'react-redux-form';

//Instruments
import { api } from '../../../../REST';
import { tasksActions } from '../../actions';
import { uiActions } from '../../../ui/actions';

export function* createTask ({ payload: message }) {
    try {
        yield put(uiActions.startSpinning());
        const response = yield apply(api, api.tasks.create, [message]);
        const { data: task, errorMessage } = yield apply(response, response.json);

        if (response.status !== 200) {
            throw new Error(errorMessage);
        }

        yield put(tasksActions.createTask(task));
        yield put(actions.reset('form.scheduler'));
    } catch (error) {
        yield put(uiActions.emitError(error, 'createTask worker'));
    } finally {
        yield put(uiActions.stopSpinning());
    }
}
