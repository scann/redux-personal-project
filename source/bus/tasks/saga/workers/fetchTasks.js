//Core
import { put, apply } from 'redux-saga/effects';

//Instruments
import { api } from '../../../../REST';
import { tasksActions } from '../../actions';

export function* fetchTasks () {
    try {
        //yield put(uiActions.startFetching());
        const response = yield apply(api, api.tasks.fetch);
        const { data: tasks, errorMessage } = yield apply(response, response.json);

        if (response.status !== 200) {
            throw new Error(errorMessage);
        }

        yield put(tasksActions.fillTasks(tasks));
    } catch (error) {
        console.log('fetchTasks worker error');
        //yield put(uiActions.emitError(error, 'fetchTasks worker'));
    } finally {
        //yield put(uiActions.stopFetching());
    }
}
