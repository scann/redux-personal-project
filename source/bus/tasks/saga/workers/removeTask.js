//Core
import { put, apply } from 'redux-saga/effects';

//Instruments
import { api } from '../../../../REST';
import { tasksActions } from '../../actions';
import { uiActions } from '../../../ui/actions';

export function* removeTask ({ payload: taskId }) {
    try {
        yield put(uiActions.startSpinning());
        const response = yield apply(api, api.tasks.remove, [taskId]);

        if (response.status !== 204) {
            const { errorMessage } = yield apply(response, response.json);

            throw new Error(errorMessage);
        }

        yield put(tasksActions.removeTask(taskId));
    } catch (error) {
        yield put(uiActions.emitError(error, 'removeTask worker'));
    } finally {
        yield put(uiActions.stopSpinning());
    }
}
