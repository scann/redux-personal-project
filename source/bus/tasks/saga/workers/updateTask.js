//Core
import { put, apply } from 'redux-saga/effects';

//Instruments
import { api } from '../../../../REST';
import { tasksActions } from '../../actions';
import { uiActions } from '../../../ui/actions';

export function* updateTask ({ payload: task }) {
    try {
        yield put(uiActions.startSpinning());
        const response = yield apply(api, api.tasks.update, [task]);

        const { data: [updatedTask], message: errorMessage } = yield apply(response, response.json);

        if (response.status !== 200) {
            throw new Error(errorMessage);
        }

        yield put(tasksActions.updateTask(updatedTask));
    } catch (error) {
        yield put(uiActions.emitError(error, 'updateTask worker'));
    } finally {
        yield put(uiActions.stopSpinning());
    }
}