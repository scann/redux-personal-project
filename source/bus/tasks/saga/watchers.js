//Core
import { takeEvery, all, call } from 'redux-saga/effects';

//Types
import { types } from '../types';

//Workers
import { fetchTasks } from './workers';

export function* watchFetchTasks () {
    yield takeEvery(types.FETCH_TASKS_ASYNC, fetchTasks);
}

export function* watchTasks () {
    yield all([call(watchFetchTasks)]);
}
