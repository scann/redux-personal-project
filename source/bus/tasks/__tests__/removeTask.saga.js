//Core
import { put, apply } from 'redux-saga/effects';
import { cloneableGenerator } from 'redux-saga/utils';

//Instruments
import { api } from '../../../REST';
import { uiActions } from '../../ui/actions';
import { tasksActions } from '../actions';
import { removeTask } from '../saga/workers';

const removeTaskAction = tasksActions.removeTaskAsync(__.testMessage.id);

const saga = cloneableGenerator(removeTask)(removeTaskAction);
let clone = null;

describe('removeTask saga', () => {
    describe('should pass until response received', () => {
        test('should dispatch "startSpinning" action', () => {
            expect(saga.next().value).toEqual(put(uiActions.startSpinning()));
        });

        test('should call a fetch request', () => {
            expect(saga.next().value).toEqual(
                apply(api, api.tasks.remove, [__.testMessage.id])
            );
            clone = saga.clone();
        });
    });

    describe('should handle a 400 status response', () => {
        test('a fetch request should return 400 status response', () => {
            expect(clone.next(__.fetchResponseFail400).value).toEqual(
                apply(__.fetchResponseFail400, __.fetchResponseFail400.json)
            );
        });

        test('should contain a response data object', () => {
            expect(clone.next(__.responseDataFail).value).toEqual(
                put(uiActions.emitError(__.emitError, 'removeTask worker'))
            );
        });

        test('should dispatch "stopSpinning" action', () => {
            expect(clone.next().value).toEqual(put(uiActions.stopSpinning()));
        });

        test('should finish', () => {
            expect(clone.next().done).toBe(true);
        });
    });

    describe('should handle a 204 status response', () => {

        test('should dispatch "removeTask" action', () => {
            expect(saga.next(__.fetchResponseSuccess204).value).toMatchSnapshot();
        });

        test('should dispatch "stopSpinning" action', () => {
            expect(saga.next().value).toMatchSnapshot();
        });

        test('should finish', () => {
            expect(saga.next().done).toBe(true);
        });
    });
});
