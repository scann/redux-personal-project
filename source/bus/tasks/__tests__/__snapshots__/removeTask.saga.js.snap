// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeTask saga should handle a 204 status response should dispatch "removeTask" action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": "TEST_ID",
      "type": "REMOVE_TASK",
    },
    "channel": null,
  },
}
`;

exports[`removeTask saga should handle a 204 status response should dispatch "stopSpinning" action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "type": "STOP_SPINNING",
    },
    "channel": null,
  },
}
`;
